ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/AtomicActionRecoveryModule.java

```diff
@@ -223,17 +230,40 @@ private void processTransactionsStatus()
                 try {
                     if (_recoveryStore.currentState(currentUid,
                             _transactionType) != StateStatus.OS_UNKNOWN) {
-                        doRecoverTransaction(currentUid);
+                        RecoverAtomicAction rcvAtomicAction = doRecoverTransaction(currentUid);
+
+                        /*
+                         * If the current AtomicAction has been recovered, its StateStatus should be OS_UNKNOWN.
+                         * If that is not the case, it means that the current AtomicAction still needs to be recovered
+                         */
+                        if (_recoveryStore.currentState(currentUid, _transactionType) != StateStatus.OS_UNKNOWN) {
```

Comment:
Two possibilities came to mind but there are other ways of determining whether recovery was successful, such as:

in  AARM processTransactionsStatus()

```
if (ActionManager.manager().get(rcvAtomicAction.get_uid()) != null) {
```

or in RecoverAtomicAction

```
   @Override
   protected void updateState () {
       recovered = super.updateState();
   }
```
etc

Reply:
Regarding the options, ActionManager has no record of the transactions once they are out of the control of the transaction manager so the first option doesn't seem like it would work? For the second option I don't think that we can do that because changing the underlying updateState to return a boolean to tell if the result was that the state was removed would somehow change the meaning of updateState a bit too much.
But I do think we should do something a bit like the second option in that we should check the in-memory state of the BasicAction lists and look for situations where we can determine that there are no resources that have heuristics and no resources that have failed, and deciding from the state of these lists (and possibly other state) whether there is still work to be done.